### PR TITLE
GH-123516: Improve JIT memory consumption by invalidating cold executors

### PR DESCRIPTION
**Pull request by savannahostrowski**
This PR succeeds #123402 and reworks the approach to use the eval breaker for the invalidation call instead of executor creation or gc (thanks @markshannon!). In experimenting, I tried a couple of different thresholds of 10k, 100k, and 1 million runs. The benchmarks for [100k](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20240923-3.14.0a0-17ece50-JIT) and [1 million](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20240923-3.14.0a0-062c54f-JIT) were most promising. Here are some relevant stats for quick reference:

**100k**
**-  [-2.4% memory](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240923-3.14.0a0-17ece50-JIT/bm-20240923-linux-x86_64-savannahostrowski-jit_inv_mem_100k-3.14.0a0-17ece50-vs-base-mem.svg)**
**- [Roughly the same performance-wise](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240923-3.14.0a0-17ece50-JIT/bm-20240923-linux-x86_64-savannahostrowski-jit_inv_mem_100k-3.14.0a0-17ece50-vs-base.svg)**

**1 million**
- [-1.5% memory](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240923-3.14.0a0-062c54f-JIT/bm-20240923-linux-x86_64-savannahostrowski-jit_inv_mem_1m-3.14.0a0-062c54f-vs-base-mem.svg)
- [Roughly the same performance-wise](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240923-3.14.0a0-062c54f-JIT/bm-20240923-linux-x86_64-savannahostrowski-jit_inv_mem_1m-3.14.0a0-062c54f-vs-base.svg)

After chatting with @brandtbucher, I've opted to open this PR with the 100k threshold. One thing to note is that we are potentially a little too liberal in invalidating executors with this threshold, but with the lack of movement in performance and a more substantial decrease in memory usage, it seemed justified. We can continue to iterate here and consider making this tunable in the future.

* Issue: gh-123516

### CODE REVIEW

File: Include/internal/pycore_interp.h [hunk 1]
```diff
@@ -261,7 +261,7 @@ struct _is {
     struct callable_cache callable_cache;
     _PyOptimizerObject *optimizer;
     _PyExecutorObject *executor_list_head;
-
+    size_t run_counter;
```
Comment (brandtbucher) [inline Include/internal/pycore_interp.h]: I'd rename this to be more descriptive. Maybe `trace_run_counter`?

File: Include/internal/pycore_optimizer.h [hunk 2]
```diff
@@ -31,7 +31,8 @@ typedef struct {
     uint8_t oparg;
     uint16_t valid:1;
     uint16_t linked:1;
-    uint16_t chain_depth:14;  // Must be big engough for MAX_CHAIN_DEPTH - 1.
+    uint16_t chain_depth:6;  // Must be big enough for MAX_CHAIN_DEPTH - 1.
```
Comment (brandtbucher) [inline Include/internal/pycore_optimizer.h]: Can you make all three of these bitfields `uint8_t`?

File: Python/ceval_gil.c [hunk 3]
```diff
@@ -1289,6 +1289,11 @@ _Py_HandlePending(PyThreadState *tstate)
         _Py_RunGC(tstate);
     }
 
+    if((breaker & _PY_EVAL_JIT_INVALIDATE_COLD_BIT) != 0) {
```
Comment (brandtbucher) [inline Python/ceval_gil.c]:
```suggestion
    if ((breaker & _PY_EVAL_JIT_INVALIDATE_COLD_BIT) != 0) {
```

File: Python/ceval_gil.c [hunk 4]
```diff
@@ -1289,6 +1289,11 @@ _Py_HandlePending(PyThreadState *tstate)
...
+        _Py_unset_eval_breaker_bit(tstate, _PY_EVAL_JIT_INVALIDATE_COLD_BIT);
+        _Py_Executors_InvalidateCold(tstate->interp);
```
Comment (brandtbucher) [inline Python/ceval_gil.c:1294]: Just a thought I had while reading through... I don't think any of the stuff that manipulates the linked list of executors is thread-safe currently. Probably not a problem for this PR, but in the future we'll probably want to go through and add `_PyEval_StopTheWorld` and `_PyEval_StartTheWorld` calls in `optimize.c`.

File: Python/optimizer.c [hunk 5]
```diff
@@ -182,6 +182,7 @@ _PyOptimizer_Optimize(
     if (err <= 0) {
         return err;
     }
+
```
Comment (brandtbucher) [inline Python/optimizer.c]:
```suggestion
```

File: Python/bytecodes.c [hunk 6]
```diff
@@ -4831,6 +4831,14 @@ dummy_func(
             assert(((_PyExecutorObject *)executor)->vm_data.valid);
         }
 
+        tier2 op(_MAKE_WARM, (--)) {
+            current_executor->vm_data.warm = true;
+            if (++tstate->interp->run_counter > 100000) {
```
Comment (brandtbucher) [inline Python/bytecodes.c]: Let's `#define` a nice name for this threshold in `pycore_optimizer.h`.

**Review by brandtbucher — COMMENTED** [review]
This is a nice change, thanks!

**Review by brandtbucher — APPROVED** [review]

File: Python/optimizer.c [hunk 7]
```diff
@@ -1659,4 +1663,42 @@ _Py_Executors_InvalidateAll(PyInterpreterState *interp, int is_invalidation)
...
+    /* Clearing an executor can deallocate others, so we need to make a list of
+     * executors to invalidate first */
+    for (_PyExecutorObject *exec = interp->executor_list_head; exec != NULL;) {
+        assert(exec->vm_data.valid);
+        _PyExecutorObject *next = exec->vm_data.links.next;
+
+        if (!exec->vm_data.warm && PyList_Append(invalidate, (PyObject *)exec) < 0) {
+                goto error;
+        }
```
Comment (picnixz) [inline Python/optimizer.c]:
```suggestion
        if (!exec->vm_data.warm && PyList_Append(invalidate, (PyObject *)exec) < 0) {
            goto error;
        }
```

**Pull request merged**